File: clients/ios/Lextures/Core/Routing/DeepLinkRouter.swift
import Foundation

/// Parsed navigation target from a push tap, universal link, or in-app notification action URL.
enum SettingsDeepLinkSection: Equatable {
    case account
    case notifications
    case learnerProfile
    /// MOB.3 Settings/Admin hub.
    case adminHub
    /// MOB.3 Audit log page inside the admin hub.
    case auditLog
}

enum DeepLinkDestination: Equatable {
    case home
    case inbox
    case review
    case insights
    case billing
    case credentials
    case checkoutSuccess(courseId: String?)
    case checkoutCancel
    case coursesList
    case settings(SettingsDeepLinkSection)
    case course(code: String, section: CourseDeepLinkSection?, itemId: String?)
    case parent(studentId: String?, section: ParentDeepLinkSection)
    /// Public board share link (`/board-links/{token}`).
    case boardLink(token: String)
    /// Live quiz join / play (`/play` or `/play/{code}`).
    case liveQuizPlay(code: String?)
}

enum ParentDeepLinkSection: Equatable {
    case dashboard
    case grades
    case attendance
    case conferences
    case notificationPrefs
}

enum CourseDeepLinkSection: String, Equatable {
    case overview
    case modules
    case grades
    case officeHours
    case feed
    case discussions
    case live
    case files
    case attendance
    case people
    case evaluations
    case library
    case groups
    case collabDocs
    case boards
    case liveQuizzes
    case behavior
    case hallPass
    case insights
}

/// Maps web-style action URLs and `lextures://` links to native navigation intents.
enum DeepLinkRouter {
    static func resolve(_ raw: String?) -> DeepLinkDestination {
        guard let raw, !raw.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return .home
        }
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if let checkout = resolveCheckout(from: trimmed) {
            return checkout
        }
        if let parent = resolveParent(from: trimmed) {
            return parent
        }
        if let path = extractPath(from: trimmed) {
            return resolvePath(path)
        }
        return .home
    }

    private static func extractPath(from value: String) -> String? {
        if value.hasPrefix("lextures://") {
            let stripped = String(value.dropFirst("lextures://".count))
            return stripped.hasPrefix("/") ? stripped : "/\(stripped)"
        }
        if value.hasPrefix("/") {
            return value
        }
        if let url = URL(string: value), let host = url.host?.lowercased() {
            if host == "lextures.com" || host.hasSuffix(".lextures.com") || host == "localhost" {
                var path = url.path
                if !path.hasPrefix("/") { path = "/\(path)" }
                return path
            }
        }
        return nil
    }

    private static func resolvePath(_ path: String) -> DeepLinkDestination {
        let segments = path.split(separator: "/").map(String.init)
        guard let first = segments.first?.lowercased(), first == "courses", segments.count >= 2 else {
            return resolveNonCoursePath(segments)
        }
        return resolveCoursePath(segments)
    }

    private static func resolveCheckout(from raw: String) -> DeepLinkDestination? {
        let urlString = raw.hasPrefix("/") ? "https://lextures.com\(raw)" : raw
        guard let url = URL(string: urlString),
              let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return nil
        }
        switch components.path {
        case "/checkout/success":
            let courseId = components.queryItems?.first(where: { $0.name == "course_id" })?.value
            return .checkoutSuccess(courseId: courseId)
        case "/checkout/cancel":
            return .checkoutCancel
        case "/me/billing":
            return .billing
        case "/me/credentials":
            return .credentials
        default:
            return nil
        }
    }

    private static func resolveParent(from raw: String) -> DeepLinkDestination? {
        let urlString = raw.hasPrefix("/") ? "https://lextures.com\(raw)" : raw
        guard let url = URL(string: urlString),
              let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return nil
        }
        let path = components.path
        guard path == "/parent" || path.hasPrefix("/parent/") else { return nil }
        let studentId = components.queryItems?.first(where: { $0.name == "student" })?.value
        let section: ParentDeepLinkSection
        if path.contains("conferences") {
            section = .conferences
        } else if path.contains("notification") {
            section = .notificationPrefs
        } else if path.contains("grades") {
            section = .grades
        } else if path.contains("attendance") {
            section = .attendance
        } else {
            section = .dashboard
        }
        return .parent(studentId: studentId, section: section)
    }

    private static func resolveNonCoursePath(_ segments: [String]) -> DeepLinkDestination {
        switch segments.first?.lowercased() {
        case "inbox":
            return .inbox
        case "review":
            return .review
        case "courses":
            return .coursesList
        case "settings":
            return resolveSettingsDeepLink(segments)
        case "parent":
            return resolveParentDeepLink(segments)
        case "board-links":
            if segments.count >= 2 {
                let token = segments[1].trimmingCharacters(in: .whitespacesAndNewlines)
                if !token.isEmpty {
                    return .boardLink(token: token)
                }
            }
            return .home
        case "play":
            let code = segments.count >= 2
                ? segments[1].trimmingCharacters(in: .whitespacesAndNewlines)
                : nil
            return .liveQuizPlay(code: (code?.isEmpty == false) ? code : nil)
        default:
            return resolveMeDeepLink(segments)
        }
    }

    private static func resolveSettingsDeepLink(_ segments: [String]) -> DeepLinkDestination {
        guard segments.count >= 2 else { return .settings(.account) }
        switch segments[1].lowercased() {
        case "account":
            return .settings(.account)
        case "notifications":
            return .settings(.notifications)
        case "learner-profile":
            return .settings(.learnerProfile)
        case "admin", "admin-console":
            if segments.count >= 3, segments[2].lowercased() == "audit-log" {
                return .settings(.auditLog)
            }
            return .settings(.adminHub)
        case "audit-log":
            return .settings(.auditLog)
        default:
            return .settings(.account)
        }
    }

    private static func resolveParentDeepLink(_ segments: [String]) -> DeepLinkDestination {
        guard segments.count >= 2 else { return .parent(studentId: nil, section: .dashboard) }
        switch segments[1].lowercased() {
        case "conferences":
            return .parent(studentId: nil, section: .conferences)
        case "notification-prefs":
            return .parent(studentId: nil, section: .notificationPrefs)
        case "grades":
            return .parent(studentId: nil, section: .grades)
        case "attendance":
            return .parent(studentId: nil, section: .attendance)
        default:
            return .parent(studentId: nil, section: .dashboard)
        }
    }

    private static func resolveMeDeepLink(_ segments: [String]) -> DeepLinkDestination {
        guard segments.count >= 2, segments[0].lowercased() == "me" else { return .home }
        switch segments[1].lowercased() {
        case "study-insights":
            return .insights
        case "credentials":
            return .credentials
        default:
            return .home
        }
    }

    private static func resolveCoursePath(_ segments: [String]) -> DeepLinkDestination {
        let courseCode = segments[1]
        guard segments.count > 2 else {
            return .course(code: courseCode, section: .overview, itemId: nil)
        }
        return resolveCourseSection(courseCode: courseCode, segments: segments)
    }

    private static func resolveCourseSection(courseCode: String, segments: [String]) -> DeepLinkDestination {
        switch segments[2].lowercased() {
        case "discussions":
            let itemId = segments.count >= 5 && segments[3].lowercased() == "threads" ? segments[4] : nil
            return .course(code: courseCode, section: .discussions, itemId: itemId)
        case "collab-docs":
            let itemId = segments.count >= 4 ? segments[3] : nil
            return .course(code: courseCode, section: .collabDocs, itemId: itemId)
        case "boards":
            let itemId = segments.count >= 4 ? segments[3] : nil
            return .course(code: courseCode, section: .boards, itemId: itemId)
        case "live-quizzes":
            return .course(code: courseCode, section: .liveQuizzes, itemId: nil)
        case "assignments", "quizzes", "modules":
            if segments.count >= 5,
               ["content", "quiz", "assignment"].contains(segments[3].lowercased()) {
                return .course(code: courseCode, section: .modules, itemId: segments[4])
            }
            let itemId = segments.count >= 4 ? segments[3] : nil
            return .course(code: courseCode, section: .modules, itemId: itemId)
        case "gradebook":
            return .course(code: courseCode, section: .grades, itemId: nil)
        default:
            let section = simpleCourseSection(for: segments[2].lowercased())
            return .course(code: courseCode, section: section, itemId: nil)
        }
    }

    private static func simpleCourseSection(for key: String) -> CourseDeepLinkSection {
        switch key {
        case "grades":
            return .grades
        case "office-hours":
            return .officeHours
        case "feed":
            return .feed
        case "live", "live-sessions":
            return .live
        case "files":
            return .files
        case "attendance":
            return .attendance
        case "people", "enrollments":
            return .people
        case "evaluations", "evaluation-results":
            return .evaluations
        case "library", "reading-dashboard":
            return .library
        case "groups":
            return .groups
        case "behavior":
            return .behavior
        case "hall-pass":
            return .hallPass
        case "insights", "at-risk", "whats-working":
            return .insights
        default:
            return .overview
        }
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/routing/DeepLinkRouter.kt
package com.lextures.android.core.routing

enum class SettingsDeepLinkSection {
    Account,
    Notifications,
    LearnerProfile,
    /** MOB.3 Settings/Admin hub. */
    AdminHub,
    /** MOB.3 Audit log page inside the admin hub. */
    AuditLog,
}

/** Parsed navigation target from a push tap, app link, or in-app notification action URL. */
sealed class DeepLinkDestination {
    data object Home : DeepLinkDestination()
    data object Inbox : DeepLinkDestination()
    data object Review : DeepLinkDestination()
    data object Insights : DeepLinkDestination()
    data object Billing : DeepLinkDestination()
    data object Credentials : DeepLinkDestination()
    data class CheckoutSuccess(val courseId: String? = null) : DeepLinkDestination()
    data object CheckoutCancel : DeepLinkDestination()
    data object CoursesList : DeepLinkDestination()
    data class Settings(val section: SettingsDeepLinkSection) : DeepLinkDestination()
    data class Course(
        val code: String,
        val section: CourseDeepLinkSection? = null,
        val itemId: String? = null,
    ) : DeepLinkDestination()

    data class Parent(
        val studentId: String? = null,
        val section: ParentDeepLinkSection = ParentDeepLinkSection.Dashboard,
    ) : DeepLinkDestination()

    /** Public board share link (`/board-links/{token}`). */
    data class BoardLink(val token: String) : DeepLinkDestination()
    /** Live quiz join / play (`/play` or `/play/{code}`). */
    data class LiveQuizPlay(val code: String? = null) : DeepLinkDestination()
}

enum class ParentDeepLinkSection {
    Dashboard,
    Grades,
    Attendance,
    Conferences,
    NotificationPrefs,
}

enum class CourseDeepLinkSection {
    Overview,
    Modules,
    Grades,
    Feed,
    Discussions,
    OfficeHours,
    Live,
    Files,
    Attendance,
    People,
    Evaluations,
    Library,
    Groups,
    CollabDocs,
    Boards,
    LiveQuizzes,
    Behavior,
    HallPass,
    Insights,
}

/** Maps web-style action URLs and `lextures://` links to native navigation intents. */
object DeepLinkRouter {
    fun resolve(raw: String?): DeepLinkDestination {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.isEmpty()) return DeepLinkDestination.Home
        resolveCheckout(trimmed)?.let { return it }
        resolveParent(trimmed)?.let { return it }
        val path = extractPath(trimmed) ?: return DeepLinkDestination.Home
        return resolvePath(path, trimmed)
    }

    private fun extractPath(value: String): String? {
        if (value.startsWith("lextures://")) {
            val stripped = value.removePrefix("lextures://")
            return if (stripped.startsWith("/")) stripped else "/$stripped"
        }
        if (value.startsWith("/")) return value
        if (value.startsWith("http://") || value.startsWith("https://")) {
            val uri = runCatching { java.net.URI(value) }.getOrNull() ?: return null
            val host = uri.host?.lowercase().orEmpty()
            if (host == "lextures.com" || host.endsWith(".lextures.com") || host == "localhost") {
                var path = uri.path.orEmpty()
                if (!path.startsWith("/")) path = "/$path"
                return path.ifEmpty { null }
            }
            return null
        }
        val uri = runCatching { android.net.Uri.parse(value) }.getOrNull() ?: return null
        val host = uri.host?.lowercase().orEmpty()
        if (host == "lextures.com" || host.endsWith(".lextures.com") || host == "localhost") {
            var path = uri.path.orEmpty()
            if (!path.startsWith("/")) path = "/$path"
            return path
        }
        return null
    }

    private fun resolveParent(raw: String): DeepLinkDestination? {
        val urlString = if (raw.startsWith("/")) "https://lextures.com$raw" else raw
        val uri = runCatching { java.net.URI(urlString) }.getOrNull() ?: return null
        val path = uri.path.orEmpty()
        if (path != "/parent" && !path.startsWith("/parent/")) return null
        val studentId = uri.rawQuery.orEmpty().split("&")
            .map { it.split("=", limit = 2) }
            .firstOrNull { it.getOrNull(0) == "student" }
            ?.getOrNull(1)
        val section = when {
            path.contains("conferences") -> ParentDeepLinkSection.Conferences
            path.contains("notification") -> ParentDeepLinkSection.NotificationPrefs
            path.contains("grades") -> ParentDeepLinkSection.Grades
            path.contains("attendance") -> ParentDeepLinkSection.Attendance
            else -> ParentDeepLinkSection.Dashboard
        }
        return DeepLinkDestination.Parent(studentId = studentId, section = section)
    }

    private fun resolvePath(path: String, raw: String = path): DeepLinkDestination {
        val segments = path.trim('/').split('/').filter { it.isNotEmpty() }
        if (segments.firstOrNull()?.lowercase() != "courses" || segments.size < 2) {
            return when {
                segments.firstOrNull()?.lowercase() == "inbox" -> DeepLinkDestination.Inbox
                segments.firstOrNull()?.lowercase() == "review" -> DeepLinkDestination.Review
                segments.firstOrNull()?.lowercase() == "courses" -> DeepLinkDestination.CoursesList
                segments.firstOrNull()?.lowercase() == "settings" -> {
                    val section = when (segments.getOrNull(1)?.lowercase()) {
                        "notifications" -> SettingsDeepLinkSection.Notifications
                        "learner-profile" -> SettingsDeepLinkSection.LearnerProfile
                        "admin", "admin-console" -> {
                            if (segments.getOrNull(2)?.lowercase() == "audit-log") {
                                SettingsDeepLinkSection.AuditLog
                            } else {
                                SettingsDeepLinkSection.AdminHub
                            }
                        }
                        "audit-log" -> SettingsDeepLinkSection.AuditLog
                        else -> SettingsDeepLinkSection.Account
                    }
                    DeepLinkDestination.Settings(section)
                }
                segments.firstOrNull()?.lowercase() == "parent" -> {
                    val studentId = runCatching { android.net.Uri.parse(raw) }.getOrNull()?.getQueryParameter("student")
                    val section = when (segments.getOrNull(1)?.lowercase()) {
                        "conferences" -> ParentDeepLinkSection.Conferences
                        "notification-prefs" -> ParentDeepLinkSection.NotificationPrefs
                        "grades" -> ParentDeepLinkSection.Grades
                        "attendance" -> ParentDeepLinkSection.Attendance
                        else -> ParentDeepLinkSection.Dashboard
                    }
                    DeepLinkDestination.Parent(studentId = studentId, section = section)
                }
                segments.size >= 2 && segments[0].equals("me", ignoreCase = true) -> when {
                    segments[1].equals("study-insights", ignoreCase = true) -> DeepLinkDestination.Insights
                    segments[1].equals("credentials", ignoreCase = true) -> DeepLinkDestination.Credentials
                    else -> DeepLinkDestination.Home
                }
                segments.firstOrNull()?.equals("board-links", ignoreCase = true) == true -> {
                    val token = segments.getOrNull(1)?.trim().orEmpty()
                    if (token.isNotEmpty()) DeepLinkDestination.BoardLink(token) else DeepLinkDestination.Home
                }
                segments.firstOrNull()?.equals("play", ignoreCase = true) == true -> {
                    val code = segments.getOrNull(1)?.trim()?.takeIf { it.isNotEmpty() }
                    DeepLinkDestination.LiveQuizPlay(code)
                }
                else -> DeepLinkDestination.Home
            }
        }

        val courseCode = segments[1]
        if (segments.size == 2) {
            return DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Overview)
        }

        return when (segments[2].lowercase()) {
            "grades" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Grades)
            "office-hours" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.OfficeHours)
            "feed" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Feed)
            "discussions" -> {
                if (segments.size >= 5 && segments[3].equals("threads", ignoreCase = true)) {
                    DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Discussions, segments[4])
                } else {
                    DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Discussions)
                }
            }
            "live", "live-sessions" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Live)
            "files" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Files)
            "attendance" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Attendance)
            "people", "enrollments" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.People)
            "evaluations", "evaluation-results" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Evaluations)
            "library", "reading-dashboard" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Library)
            "groups" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Groups)
            "behavior" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Behavior)
            "hall-pass" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.HallPass)
            "insights", "at-risk", "whats-working" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Insights)
            "collab-docs" -> DeepLinkDestination.Course(
                code = courseCode,
                section = CourseDeepLinkSection.CollabDocs,
                itemId = segments.getOrNull(3),
            )
            "boards" -> DeepLinkDestination.Course(
                code = courseCode,
                section = CourseDeepLinkSection.Boards,
                itemId = segments.getOrNull(3),
            )
            "live-quizzes" -> DeepLinkDestination.Course(
                code = courseCode,
                section = CourseDeepLinkSection.LiveQuizzes,
            )
            "gradebook" -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Grades)
            "assignments", "quizzes", "modules" -> {
                if (segments.size >= 5 && segments[3].lowercase() in setOf("content", "quiz", "assignment")) {
                    DeepLinkDestination.Course(
                        code = courseCode,
                        section = CourseDeepLinkSection.Modules,
                        itemId = segments[4],
                    )
                } else {
                    DeepLinkDestination.Course(
                        code = courseCode,
                        section = CourseDeepLinkSection.Modules,
                        itemId = segments.getOrNull(3),
                    )
                }
            }
            else -> DeepLinkDestination.Course(courseCode, CourseDeepLinkSection.Overview)
        }
    }

    private fun resolveCheckout(raw: String): DeepLinkDestination? {
        val urlString = if (raw.startsWith("/")) "https://lextures.com$raw" else raw
        val uri = runCatching { java.net.URI(urlString) }.getOrNull() ?: return null
        val query = uri.rawQuery.orEmpty()
        return when (uri.path) {
            "/checkout/success" -> {
                val courseId = query.split("&")
                    .map { it.split("=", limit = 2) }
                    .firstOrNull { it.getOrNull(0) == "course_id" }
                    ?.getOrNull(1)
                DeepLinkDestination.CheckoutSuccess(courseId)
            }
            "/checkout/cancel" -> DeepLinkDestination.CheckoutCancel
            "/me/billing" -> DeepLinkDestination.Billing
            "/me/credentials" -> DeepLinkDestination.Credentials
            else -> null
        }
    }
}
